src/main/java/org/jenkinsci/plugins/workflow/multibranch/ResolveScmStep.java

```diff
@@ -0,0 +1,297 @@
+import hudson.Extension;
+import hudson.model.Descriptor;
+import hudson.model.TaskListener;
+import hudson.scm.SCM;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMRevision;
+import jenkins.scm.api.SCMSource;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.workflow.steps.AbstractStepDescriptorImpl;
+import org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution;
+import org.jenkinsci.plugins.workflow.steps.Step;
+import org.jenkinsci.plugins.workflow.steps.StepContext;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+
+/**
+ * Resolves an {@link SCM} from a {@link SCMSource} using a priority list of target branch names.
+ *
+ * @since FIXME
+ */
+public class ResolveScmStep extends Step {
+
+    /**
+     * The {@link SCMSource}
+     */
+    @NonNull
+    private final SCMSource source;
+
+    /**
+     * The {@link SCMSource}
+     */
+    @NonNull
+    private final List<String> targets;
+
+    /**
+     * If {@code true} then {@code null} will be returned in the event that none of the target branch names can be
+     * resolved.
+     */
+    private boolean ignoreErrors;
+
+    /**
+     * Constructor.
+     *
+     * @param source  The {@link SCMSource}
+     * @param targets The {@link SCMSource}
+     */
+    @DataBoundConstructor
+    public ResolveScmStep(@NonNull SCMSource source, @NonNull List<String> targets) {
+        this.source = source;
+        this.targets = new ArrayList<String>(targets);
+    }
+
+    /**
+     * Gets the {@link SCMSource} to resolve from.
+     *
+     * @return the {@link SCMSource} to resolve from.
+     */
+    @NonNull
+    public SCMSource getSource() {
+        return source;
+    }
+
+    /**
+     * Gets the {@link SCMHead} names to try and resolve.
+     *
+     * @return the {@link SCMHead} names to try and resolve.
+     */
+    @NonNull
+    public List<String> getTargets() {
+        return Collections.unmodifiableList(targets);
+    }
+
+    /**
+     * Returns {@code true} if and only if errors will be ignored.
+     *
+     * @return {@code true} if and only if errors will be ignored.
+     */
+    public boolean isIgnoreErrors() {
+        return ignoreErrors;
+    }
+
+    /**
+     * Sets the error handling behaviour.
+     *
+     * @param ignoreErrors {@code true} if and only if errors will be ignored.
+     */
+    @DataBoundSetter
+    public void setIgnoreErrors(boolean ignoreErrors) {
+        this.ignoreErrors = ignoreErrors;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(context, this);
```

Comment:
Do not mix Guiceless `Step` with Guiceful `AbstractStep{Execution,Descriptor}Impl`.